File: wecode/1-greedy-method/Bai1.cpp
#include<iostream>
#include<stack>
#include<string>
using namespace std;

string deleteKdigits(string input,int k){
    int len = input.size();
    if(k>len){
        return "error";
    }
    else if(k==len){
        return "0";
    }
    else{
        stack<char> st;
        // iterate through input number
        for(char i:input){
            // sequencially delete bigger number
            // save small number to stack
            while(!st.empty()&&i<st.top()&&k>0){
                st.pop();
                k--;
            }
            if( !st.empty() || i!='0'){
                st.push(i);
            }
        }
        // remove largest value on top of the stack
        while (!st.empty() && k--)
		    st.pop();
	    if (st.empty())
		    return "0";

        while(!st.empty()){
            input[len-1]=st.top();
            st.pop();
            len--;
        }
    }
    return input.substr(len);
}

int main(){
    string input;
    int k;
    getline(cin,input);
    cin>>k;
    cout<<deleteKdigits(input,k);
}

File: wecode/1-greedy-method/Bai1.py
def removeKdigits(num, k):
 
    n = len(num)
    mystack = []
     
    # Store the final string in stack
    for c in num:
        while (len(mystack) > 0 and k > 0
               and mystack[-1] > c):
            mystack.pop()
            k -= 1
 
        if (len(mystack) > 0 or c != '0'):
            mystack.append(c)
 
    # Now remove the largest values from the top of the stack
    while (len(mystack) > 0 and k):
        k -= 1
        mystack.pop()
    if (len(mystack) == 0):
        return "0"

    while (len(mystack) > 0):
        num = num.replace(num[n - 1] , mystack[-1])
        mystack.pop()
        n -= 1
 
    return num[n:]
 
# main
def main():
    number = input()
    k = int(input())
    print(removeKdigits(number, k))

if __name__ == "__main__":
    main()
